File: cses/usaco/dsu_practice.py
import sys
class DSU:
    """Disjoint Set union class"""
    def __init__(self,vertices):
        self.parents = [i for i in range(0,vertices+1)]
    def find(self,i):
        if self.parents[i] != i:
            self.parents[i] = self.find(self.parents[i])
        return self.parents[i]
    def union(self,i,j):
        root_i = self.find(i)
        root_j = self.find(j)
        if root_i == root_j:
            return
        elif root_i < root_j:
            self.parents[j] = root_i
        else:
            self.parents[i] = root_j

input = sys.stdin.readline

N,Q = int(input().split())
dsu = DSU(N)
for _ in range(Q):
    t,v,u = int(input().split())
    if  t == 0:
        dsu.union(v,u)
    elif t==1:
        print(dsu.find(v)==dsu.find(u)) 

File: usaco/dsu_practice.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
class DSU
{
private:
    int vertices;
    vector<int> parents;

public:
    DSU(int n)
    {
        vertices = n;
        parents.resize(n);
        for (int i = 0; i < n; i++)
        {
            parents[i] = i;
        }
    }
    int find(int i)
    {
        if (parents[i] != i)
        {
            parents[i] = find(parents[i]); // path-compress
        }
        return parents[i];
    }
    void unionB(int i, int j)
    {
        int root_i = find(i);
        int root_j = find(j);
        if (root_i == root_j)
        {
            return;
        }
        else if (root_i < root_j)
        {
            parents[j] = root_i;
        }
        else
        {
            parents[i] = root_j;
        }
    }
};

int main()
{
    ll N, Q;
    cin >> N >> Q;
    DSU d(N);
    for (int i = 0; i < Q; i++)
    {
        /* code */
        ll t, a, b;
        cin >> t >> a >> b;
        if (t == 0)
        {
            d.unionB(a, b);
        }
        else if (t == 1)
        {
            cout << int(d.find(a) == d.find(b)) << endl;
        }
    }
}